include/ginkgo/core/preconditioner/ilu.hpp

```diff
@@ -131,25 +131,97 @@ class Ilu : public EnableLinOp<
         Ilu<typename USolverType::transposed_type,
             typename LSolverType::transposed_type, ReverseApply, IndexType>;
 
-    GKO_CREATE_FACTORY_PARAMETERS(parameters, Factory)
-    {
+    class Factory;
+
+    struct parameters_type
+        : public enable_parameters_type<parameters_type, Factory> {
         /**
          * Factory for the L solver
          */
-        std::shared_ptr<typename l_solver_type::Factory>
-            GKO_FACTORY_PARAMETER_SCALAR(l_solver_factory, nullptr);
+        std::shared_ptr<const typename l_solver_type::Factory>
+            l_solver_factory{};
 
         /**
          * Factory for the U solver
          */
-        std::shared_ptr<typename u_solver_type::Factory>
-            GKO_FACTORY_PARAMETER_SCALAR(u_solver_factory, nullptr);
+        std::shared_ptr<const typename u_solver_type::Factory>
+            u_solver_factory{};
 
         /**
          * Factory for the factorization
          */
-        std::shared_ptr<LinOpFactory> GKO_FACTORY_PARAMETER_SCALAR(
-            factorization_factory, nullptr);
+        std::shared_ptr<const LinOpFactory> factorization_factory{};
+
+        [[deprecated("use with_l_solver instead")]] parameters_type&
+        with_l_solver_factory(
+            deferred_factory_parameter<typename l_solver_type::Factory> solver)
+        {
+            return with_l_solver(std::move(solver));
+        }
+
+        parameters_type& with_l_solver(
```

Comment:
I think this won't work entirely, since I am also renaming the factory parameter at the same time. The variable is still named _factory for interface-stability.